lib/pgBackRest/Archive/ArchivePushFile.pm

```diff
@@ -153,34 +154,51 @@ sub archivePushFile
 
     # Check if the WAL already exists in the repo
     my ($strArchiveId, $strChecksum, $strWarning) = archivePushCheck(
-        $oFile, $strWalFile, $strDbVersion, $ullDbSysId, walIsSegment($strWalFile) ? "${strWalPath}/${strWalFile}" : undef);
+        $strWalFile, $strDbVersion, $ullDbSysId, walIsSegment($strWalFile) ? "${strWalPath}/${strWalFile}" : undef);
 
     # Only copy the WAL segment if checksum is not defined.  If checksum is defined it means that the WAL segment already exists
     # in the repository with the same checksum (else there would have been an error on checksum mismatch).
     if (!defined($strChecksum))
     {
         my $strArchiveFile = "${strArchiveId}/${strWalFile}";
 
+        # Copy the segment to a temp file
+        my $strTempFile = $strArchiveFile;
+
+        if (walIsSegment($strWalFile) && $bCompress)
+        {
+            $strTempFile .= qw{.} . COMPRESS_EXT;
+        }
+
+        $strTempFile .= qw{.} . STORAGE_TEMP_EXT;
+
+        # Open source file
+        my $oSourceFileIo = storageDb()->openRead(
+            "${strWalPath}/${strWalFile}", {rhyFilter => [{strClass => STORAGE_FILTER_SHA}]});
+
+        $oStorageRepo->copy(
+            $oSourceFileIo,
+            $oStorageRepo->openWrite(
+                STORAGE_REPO_ARCHIVE . "/${strTempFile}",
+                {rhyFilter => walIsSegment($strWalFile) && $bCompress ? [{strClass => STORAGE_FILTER_GZIP}] : undef,
+                    bPathCreate => true}));
+
+        # Append digest if this is a WAL segment
+        if (walIsSegment($strWalFile))
+        {
+            $strArchiveFile .= '-' . $oSourceFileIo->result(STORAGE_FILTER_SHA);
+        }
+
         # Append compression extension
         if (walIsSegment($strWalFile) && $bCompress)
         {
-            $strArchiveFile .= '.' . $oFile->{strCompressExtension};
+            $strArchiveFile .= '.' . COMPRESS_EXT;
         }
 
-        # Copy the WAL segment
-        $oFile->copy(
-            PATH_DB_ABSOLUTE, "${strWalPath}/${strWalFile}",        # Source type/file
-            PATH_BACKUP_ARCHIVE, $strArchiveFile,                   # Destination type/file
-            false,                                                  # Source is not compressed
-            walIsSegment($strWalFile) && $bCompress,                # Destination compress is configurable
-            undef, undef, undef,                                    # Unused params
-            true,                                                   # Create path if it does not exist
-            undef, undef,                                           # Default User and group
-            walIsSegment($strWalFile),                              # Append checksum if WAL segment
-            $bRepoSync);                                            # Sync repo directories?
+        # Move to final location
+        $oStorageRepo->move(STORAGE_REPO_ARCHIVE . "/${strTempFile}", STORAGE_REPO_ARCHIVE . "/${strArchiveFile}");
```

Comment:
This has been fixed in the s3 branch and will be pulled back to this branch soon.